src/components/tl/cuda/tl_cuda_team_topo.c

```diff
@@ -200,61 +200,97 @@ ucc_tl_cuda_team_topo_init_rings(const ucc_tl_cuda_team_t *team,
 }
 
 static ucc_status_t
-ucc_tl_cuda_team_topo_init_proxy(const ucc_tl_cuda_team_t *team,
-                                 ucc_tl_cuda_team_topo_t *topo)
+ucc_tl_cuda_team_topo_init_proxies(const ucc_tl_cuda_team_t *team,
+                                   ucc_tl_cuda_team_topo_t *topo)
 {
     ucc_rank_t size        = UCC_TL_TEAM_SIZE(team);
     ucc_rank_t num_proxies = 0;
-    ucc_rank_t i, j, p, k;
+    ucc_rank_t i, j ,p, k, proxy;
+    float *data;
+    float score, min_score;
     ucc_status_t status;
 
     for (i = 0; i < size * size; i++) {
         if (topo->matrix[i] == 0) {
-            num_proxies ++;
+            num_proxies++;
         }
     }
 
     topo->num_proxies = num_proxies;
     if (num_proxies == 0) {
         return UCC_OK;
     }
+
     topo->proxies = (ucc_tl_cuda_proxy_t*)ucc_malloc(
             num_proxies * sizeof(ucc_tl_cuda_proxy_t), "cuda_topo_proxies");
     if (!topo->proxies) {
-        tl_error(UCC_TL_TEAM_LIB(team), "failed to alloc cuda topo proxy");
+        tl_error(UCC_TL_TEAM_LIB(team), "failed to alloc cuda topo proxies");
         return UCC_ERR_NO_MEMORY;
     }
 
+    data = (float*)ucc_malloc(size * size * sizeof(float),
+                              "cuda topo proxies data");
+    if (!data) {
+        tl_error(UCC_TL_TEAM_LIB(team), "failed to alloc cuda topo work array");
+        status = UCC_ERR_NO_MEMORY;
+        goto free_proxy;
+    }
+
+    for (i = 0; i < size; i++) {
+        for (j = 0; j < size; j++) {
+            if (ucc_tl_cuda_team_topo_is_direct(&team->super, topo, i, j)) {
+                data[i * size + j] = 1.0;
+            } else {
+                data[i * size + j] = 0.0;
+            }
+        }
+    }
+
     p = 0;
     for (i = 0; i < size; i++) {
         for (j = 0; j < size; j++) {
             if (ucc_tl_cuda_team_topo_is_direct(&team->super, topo, i, j)) {
                 continue;
             }
+            proxy = UCC_RANK_INVALID;
+            min_score = (float)(UCC_RANK_MAX);
             for (k = 0; k < size; k++) {
                 if (ucc_tl_cuda_team_topo_is_direct(&team->super, topo, i, k) &&
                     ucc_tl_cuda_team_topo_is_direct(&team->super, topo, k, j)) {
-                    topo->proxies[p].src   = i;
-                    topo->proxies[p].dst   = j;
-                    topo->proxies[p].proxy = k;
-                    break;
+                    score = ucc_max((data[i * size + k] + 1.0) /
```

Comment:
ucc_assert(topo->matrix[i*size + k] > 0 && topo->matrix[k*size +j]);
i understand "topo_is_direct" guarantees the assertion above, but i'd add for readability